streaming/base/storage/upload.py

```diff
@@ -393,6 +396,89 @@ def check_bucket_exists(self, remote: str):
+            2. If ``out`` is a remote directory, a local temporary directory is created to
+               cache the shard files and then the shard files are uploaded to a remote
+               location. At the end, the temp directory is deleted once shards are uploaded.
+            3. If ``out`` is a tuple of ``(local_dir, remote_dir)``, shard files are saved in
+               the `local_dir` and also uploaded to a remote location.
+        keep_local (bool): If the dataset is uploaded, whether to keep the local dataset
+            shard file or remove it after uploading. Defaults to ``False``.
+        progress_bar (bool): Display TQDM progress bars for uploading output dataset files to
+            a remote location. Default to ``False``.
+    """
+
+    def __init__(self,
+                 out: Union[str, Tuple[str, str]],
+                 keep_local: bool = False,
+                 progress_bar: bool = False) -> None:
+        super().__init__(out, keep_local, progress_bar)
+
+        import boto3
+
+        # Create a session and use it to make our client. Unlike Resources and Sessions,
+        # clients are generally thread-safe.
+        session = boto3.session.Session()
+        self.r2_client = session.client('s3',
+                                        region_name='auto',
+                                        endpoint_url=os.environ['R2_ENDPOINT_URL'],
+                                        aws_access_key_id=os.environ['R2_ACCESS_KEY_ID'],
+                                        aws_secret_access_key=os.environ['R2_SECRET_ACCESS_KEY'])
+        self.check_bucket_exists(self.remote)  # pyright: ignore
+
+    def upload_file(self, filename: str):
+        """Upload file from local instance to Cloudflare R2 bucket.
+
+        Args:
+            filename (str): File to upload.
+        """
+        local_filename = os.path.join(self.local, filename)
+        local_filename = local_filename.replace('\\', '/')
+        remote_filename = os.path.join(self.remote, filename)  # pyright: ignore
+        remote_filename = remote_filename.replace('\\', '/')
+        obj = urllib.parse.urlparse(remote_filename)
+        logger.debug(f'Uploading to {remote_filename}')
+        file_size = os.stat(local_filename).st_size
+        with tqdm.tqdm(total=file_size,
+                       unit='B',
+                       unit_scale=True,
+                       desc=f'Uploading to {remote_filename}',
+                       disable=(not self.progress_bar)) as pbar:
+            self.r2_client.upload_file(
+                local_filename,
+                obj.netloc,
+                obj.path.lstrip('/'),
+                Callback=lambda bytes_transferred: pbar.update(bytes_transferred),
+            )
+        self.clear_local(local=local_filename)
+
+    def check_bucket_exists(self, remote: str):
+        """Raise an exception if the bucket does not exist.
+
+        Args:
+            remote (str): S3 bucket path.
```

Comment:
nit: S3 bucket path. -> `r2 bucket path.`

Reply:
Addressed as part of https://github.com/mosaicml/streaming/pull/255/commits/3b7e2068f524d6ff2a3b9de4908ff7064a8eb5eb.